source/contracts/reporting/ReportingWindow.sol

```diff
@@ -209,6 +215,10 @@ contract ReportingWindow is DelegationTarget, Typed, Initializable, IReportingWi
         return getUniverse().getReportingWindowByTimestamp(_previousTimestamp);
     }
 
+    function allMarketsFinalized() constant public returns (bool) {
+        return markets.count == finalizedMarketCount;
```

Comment:
We don't have any strong invariants that these will eventually equal each other.  As mentioned above, it would be nice if we could come up with some mechanisms that provide us strong guarantees without the reader/auditor having to read/understand everything just to assert that all reporting windows will eventually land in this state.  I don't have any great solution at the moment, though perhaps we can come up with some kind of specialized container library that gives us stronger guarantees that markets eventually end up in this state?

Reply:
What would a container library like that buy us exactly? Is the worry that a market will finalize and incorrectly not update the count?

Reply:
I'm not sure exactly what it would look like.  I just always am worried anytime there is something as large/complex as Market and it doesn't have strong invariants at every step.  I don't have a specific bug that I know of, but my gut tells me that there is _probably_ a bug somewhere in market that could result in a market not being accounted for properly.

In general, the level of complexity in market makes me uncomfortable, and denormalized data like this without strong invariants is a big source of that.

Reply:
Yea it's a complex state machine so I get the concern.

Reply:
As a note, if you think of assertions / invariants that you think should always be true in the code in certain places, add them in as `assert`s, then we can statically analyze things and actually prove that they're always true, or if not, find when they're _not_ true.